crates/loader/src/oci/cache.rs

```diff
@@ -0,0 +1,166 @@
+const DATA_DIR: &str = "data";
+
+const MANIFEST_FILE: &str = "manifest.json";
+const CONFIG_FILE: &str = "config.json";
+const LATEST_TAG: &str = "latest";
+
+/// Cache for registry entities.
+pub struct Cache {
+    /// Root directory for the cache instance.
+    pub root: PathBuf,
+}
+
+impl Cache {
+    /// Create a new cache given an optional root directory.
+    pub async fn new(root: Option<PathBuf>) -> Result<Self> {
+        let root = match root {
+            Some(root) => root,
+            None => dirs::cache_dir()
+                .context("cannot get cache directory")?
+                .join(CONFIG_DIR),
+        };
+        let root = root.join(REGISTRY_CACHE_DIR);
+        Self::ensure_dirs(&root).await?;
+
+        Ok(Self { root })
+    }
+
+    /// The manifests directory for the current cache.
+    fn manifests_dir(&self) -> PathBuf {
+        self.root.join(MANIFESTS_DIR)
+    }
+
+    /// The Wasm bytes directory for the current cache.
+    fn wasm_dir(&self) -> PathBuf {
+        self.root.join(WASM_DIR)
+    }
+
+    /// The data directory for the current cache.
+    fn data_dir(&self) -> PathBuf {
+        self.root.join(DATA_DIR)
+    }
+
+    /// Return the path to a wasm file given its digest.
+    pub fn wasm_file(&self, digest: impl AsRef<str>) -> Result<PathBuf> {
+        let path = &self.wasm_dir().join(digest.as_ref());
+        match path.exists() {
+            true => Ok(path.into()),
+            false => bail!(format!(
+                "cannot find wasm file for digest {}",
+                digest.as_ref()
+            )),
+        }
+    }
+
+    /// Return the path to a data file given its digest.
+    pub fn data_file(&self, digest: impl AsRef<str>) -> Result<PathBuf> {
+        let path = &self.data_dir().join(digest.as_ref());
+        match path.exists() {
+            true => Ok(path.into()),
+            false => bail!(format!(
+                "cannot find data file for digest {}",
+                digest.as_ref()
+            )),
+        }
+    }
+
+    /// Get the file path to an OCI manifest given a reference.
+    /// If the directory for the manifest does not exist, this will create it.
+    pub async fn manifest_for_reference(&self, reference: impl AsRef<str>) -> Result<PathBuf> {
+        let reference: Reference = reference
+            .as_ref()
+            .parse()
+            .context("cannot parse reference")?;
+        let p = self
+            .manifests_dir()
+            .join(reference.registry())
+            .join(reference.repository())
+            .join(reference.tag().unwrap_or(LATEST_TAG));
+
+        if !p.is_dir() {
+            fs::create_dir_all(&p)
+                .await
+                .context("cannot find directory for manifest")?;
+        }
+
+        Ok(p.join(MANIFEST_FILE))
+    }
+
+    /// Get the file path to the OCI configuration object given a reference.
+    pub async fn config_for_reference(&self, reference: impl AsRef<str>) -> Result<PathBuf> {
```

Comment:
Similarly I'd suggest `oci_config_file(...)`.

I am not sure quite where to draw the boundary but given that the cache knows about Wasm vs data, I wonder whether we should consider calling this `lockfile(...)` - do the layers above this care that the lockfile is implemented using the OCI config object?  (This is a question rather than a suggestion.)